Replace Cumulative implementation by state-of-the-art implementation of TimeTabling and OverloadChecking

ArthurGodet opened the pull request:
Also fixes #1114

cprudhom committed:
Prepare the code for the next version

cprudhom committed:
Update publish-snapshot.yml (#1100)

set the relative path to the settings

cprudhom committed:
Update CHANGES.md

cprudhom committed:
Bump version

cprudhom committed:
Update shade conf to generate the lite file with the right name

cprudhom committed:
Update CHANGES.md

cprudhom committed:
Add better exceptions when an unknwon operator is declared

cprudhom committed:
Remove dependencies to xchart (#1103)

cprudhom committed:
Add sweep-based DiffN

cprudhom committed:
Add "limits" as alias in RegParser

cprudhom committed:
Fix #1122

cprudhom committed:
Fix PickOnFil (incorrect API call) + fix issue related to failure in PropagationInsight

cprudhom committed:
Deal with UnsupportedOperationException instead of throwing it when SizeOf fails at estimating the size of the model.

cprudhom committed:
Add new APIs to ArrayUtils:
- interlace n arrays
- extract a sub-matrix from a matrix

cprudhom committed:
Remove `buildDistanceMatrix` and `buildDifferenceMatrix` from Solver (ie, IOutputFactory)

cprudhom committed:
Update header to 2025

cprudhom committed:
Remove lgtm and travis files + bagdes

cprudhom committed:
Bump codeQL version

cprudhom committed:
Fix incorrect indices in Flatzinc#set_diff

cprudhom committed:
Fix bug in XCSPParser#buildCtrElement

cprudhom committed:
Change the way element2d is decomposed (reduce the auxiliary variables to 1 + 2 views)

cprudhom committed:
Update Javadoc of DiffN

cprudhom committed:
Fix typo in param name (PickOnDom)

cprudhom committed:
Add wrong decisions in VerboseSolving

cprudhom committed:
Add "Group" of variables (clearer alternative to hook)

cprudhom committed:
Add InnerOuterCutoff as a new restart sequence

cprudhom committed:
Add RoundRobin strategy (for IntVar)
+ remove 'scores' structure from PickOnFil

cprudhom committed:
Fix error in commit #20e72715

cprudhom committed:
initiate release 4.10.18

cprudhom committed:
Merge branch 'release-4.10.18'

cprudhom committed:
merge dev_lcg into master

ArthurGodet committed:
replace Cumulative implementation by state-of-the-art implementation of TimeTabling and OverloadChecking

ArthurGodet committed:
replace Cumulative implementation by state-of-the-art implementation of TimeTabling and OverloadChecking

ArthurGodet committed:
fix FConstraint in choco-parsers + add header in Profile

ArthurGodet committed:
fix FConstraint in choco-parsers + add header in Profile

ArthurGodet committed:
add explanations to PropagatorCumulative

ArthurGodet committed:
add explanations to PropagatorCumulative

cprudhom commented:
I change the target branch from `master` to `develop`

cprudhom commented on solver/src/main/java/org/chocosolver/solver/constraints/nary/cumulative/Event.java:
This file could be annotated a little more.

cprudhom commented on solver/src/test/java/org/chocosolver/solver/constraints/checker/Modeler.java:
```diff
@@ -670,7 +670,7 @@ public Model model(int n, int[][] domains, THashMap<int[], IntVar> map, Object p
                 h[i] = vars[i + 3 * k];
             }
             IntVar capa = vars[vars.length - 1];
-            model.cumulative(tasks, h, capa, (boolean)parameters).post();
```
Ignoring this parameter should also involve the calling test, right?

ArthurGodet replied:
What calling test ?

cprudhom commented on solver/src/main/java/org/chocosolver/solver/constraints/nary/cumulative/EventPointSeries.java:
This file could be annotated a little more.

cprudhom commented on solver/src/main/java/org/chocosolver/solver/constraints/nary/cumulative/Profile.java:
This file could be annotated a little more.

cprudhom commented on solver/src/main/java/org/chocosolver/solver/constraints/nary/cumulative/PropagatorCapacity.java:
This file could be annotated a little more.

cprudhom reviewed (commented):
Apart from the missing comments, everything is OK.

jgFages commented on solver/src/main/java/org/chocosolver/solver/constraints/ISchedulingFactory.java:
```diff
@@ -0,0 +1,251 @@
+            keptHeights[i] = heightsToKeep.get(i);
+        }
+        if (keptTasks.length == 1) {
+            return ref().arithm(keptHeights[0], "<=", capacity);
```
height <= capa OR duration = 0, no ?

ArthurGodet replied:
Indeed, it should be this formula. Now the question is : is it better to use an OR constraint or directly the CUMULATIVE constraint on the only task ?

jgFages replied:
I would expect the OR to be more efficient

ArthurGodet replied:
I will manage it with PropagatorCapacity as it is doing exactly that (and is working with optional task)

jgFages replied:
ok

jgFages commented:
The branch needs to be rebased from origin/develop

cprudhom committed:
Ibex: migrate script and workflow to v2.9.1 (#1168)

Ibex: migrate script and workflow to v2.9.1
- update the installation steps from the ibex install script
- update pom.xml to make tests work
- change workflow to export LD_LIBRARY_PATH just before calling the test suite 
- add a dockerfile for test purpose

ArthurGodet committed:
Merge pull request #1166 from chocoteam/feature/cumulative-explained

Add explanations to PropagatorCumulative

ArthurGodet committed:
Merge branch 'develop' of https://github.com/chocoteam/choco-solver into feature/cumulative

ArthurGodet committed:
add Javadoc and some comments in Cumulative's classes

ArthurGodet committed:
add Javadoc and some comments in Cumulative's classes

ArthurGodet committed:
cumulative() declares only PropagatorCapacity if there is only one kept tasks

ArthurGodet committed:
cumulative() declares only PropagatorCapacity if there is only one kept tasks

ArthurGodet committed:
add heights to explanations of PropagatorCumulative

ArthurGodet committed:
add heights to explanations of PropagatorCumulative

ArthurGodet committed:
add explanations for PropagatorCapacity

ArthurGodet committed:
add explanations for PropagatorCapacity

ArthurGodet committed:
CumulativeTest : add RCPSP tests, JSSP tests and testPropagatorCapacity

ArthurGodet committed:
CumulativeTest : add RCPSP tests, JSSP tests and testPropagatorCapacity

ArthurGodet committed:
add PropGraphCumulative

ArthurGodet committed:
add PropGraphCumulative

ArthurGodet committed:
make Task propagate s+d=e relation as soon as possible

ArthurGodet committed:
make Task propagate s+d=e relation as soon as possible

ArthurGodet committed:
change Event's comparator to treat ECP events before SCP events at same timepoint

ArthurGodet committed:
change Event's comparator to treat ECP events before SCP events at same timepoint

ArthurGodet committed:
improve EventPointSeries wrt performed tasks

ArthurGodet committed:
improve EventPointSeries wrt performed tasks

ArthurGodet committed:
simplify code in PropagatorResource + add PropagatorResource.disjoint() method

ArthurGodet committed:
simplify code in PropagatorResource + add PropagatorResource.disjoint() method

ArthurGodet committed:
use of ref().taskVar() instead of new Task() in ISchedulingFactory

ArthurGodet committed:
use of ref().taskVar() instead of new Task() in ISchedulingFactory

ArthurGodet committed:
imrpove code in PropagatorCumulative

ArthurGodet committed:
imrpove code in PropagatorCumulative

ArthurGodet committed:
Merge branch 'feature/cumulative' of https://github.com/chocoteam/choco-solver into feature/cumulative

ArthurGodet committed:
simplify code and improve performance for the PropagatorCumulative

ArthurGodet committed:
fix problem in PropagatorCumulative.isEntailed()

ArthurGodet committed:
add energyNaive() and disjunctiveEnergyNaive() to PropagatorCumulative + update flatzinc and xcsp expected number of solutions, nodes and fails

ArthurGodet committed:
PropagatorCumulative energyNaive() and disjunctiveEnergyNaive() are used only if parameters are set to true + fix explanations in Task and PropagatorCumulative

ArthurGodet commented:
I had problems with git rebase, so, in order to keep a clean git history, we need to squash this PR whenever it is accepted/merged

jgFages commented on CHANGES.md:
```diff
@@ -70,23 +70,6 @@ See [milestone 5.0.0](https://github.com/chocoteam/choco-solver/milestone/40)
-
-**Full Changelog**: https://github.com/chocoteam/choco-solver/compare/v4.10.15...v4.10.16
-
-
```
is this change normal ?

ArthurGodet replied:
Aren't these changes written twice (hence the deleted lines) ?

jgFages commented on solver/src/main/java/org/chocosolver/solver/constraints/nary/cumulative/NRJCumulFilter.java:
```diff
@@ -1,100 +0,0 @@
-				sor_array[idx++] = i;
-			}
-		}
-		sorter.sort(sor_array,idx,comparator);
```
Is this logic implemented somewhere else ? 
If not, I would like to keep it, within the new framework

ArthurGodet replied:
Yes, it is kept in the function energyNaive() of PropagatorCumulative. It is not activated by default (we tested on instances of the MiniZinc challenge with Charles), but it is kept for the cases where it helps.

ArthurGodet replied:
However, if you talk about the while loop storing values in slb, hlb, eub arrays, then no it is not kept.
Do you see an interest in keeping it ?

jgFages replied:
Ah great thank you very much I didn't see the energyNaive()

cprudhom committed:
Explanation tests: extend framework to deal with Task

cprudhom committed:
Explanation tests: extend the framework to include the cumulative constraint

ArthurGodet committed:
merge xcsp into feature/cumulative

ArthurGodet committed:
fix Profile sweep algorithm for explanations + WIP for explanations in PropagatorCumulative

ArthurGodet committed:
fix explanations for PropagatorCumulative

ArthurGodet committed:
make sure cumulative tests run in less than 1 minute in TestExplanation

ArthurGodet commented:
Important to note : the failing tests do not rely on the cumulative constraint

ArthurGodet committed:
fix Task.updateDuration() when using lcg + simplify code for explanations in Task.propagate()

cprudhom commented:
Here are some results of the new Cumulative (`new`) in comparison to the current version (`base`) on instances from the last 6 MiniZinc challenges declaring `cumulative` constraints.

## With LCG off
<img width="512" height="550" alt="image" src="https://github.com/user-attachments/assets/3797f521-f159-4338-8f57-67445741f81f" />

The two instances that are faster with `base` than with `new` are :
- `gfd-schedule2_n200f5d50m40k6_10124` : 95s vs 100s 
- `model4_opt_test03` : 106s vs 154s

The difference comes from the usage of the OverloadChecking algorithm which never helps on these instances.

<img width="571" height="455" alt="image" src="https://github.com/user-attachments/assets/ad2090d5-d0c8-46d2-adf0-b440ebb570f4" />


## With LCG on

<img width="533" height="550" alt="image" src="https://github.com/user-attachments/assets/0ae228ef-47ef-490a-9b49-72fe49786d35" />

The two instances where `base-lcg` is faster than `new-lcg` are:
- `mrcpsp_j30_29_3` : 22s vs 57s
- `mrcpsp_j30_54_2` : 12s vs 23s

I still need to understand the differences but as a recall, `base-lcg` time-based decomposes the cumulative constraint.

<img width="571" height="455" alt="image" src="https://github.com/user-attachments/assets/3e3a68bf-086f-48db-8a48-068865441f63" />

ArthurGodet commented:
I confirm that base-lcg is based on decomposition for the cumulative. I wonder if the two instances where base-lcg does better are on a short horizon ? That might explain the performance difference.

cprudhom commented:
Actually, in the benchmarks, there are 5 instances of mrcpsp_j30, with different results.


Instance | Opt | base-lcg | new-lcg
-- | -- | -- | --
mrcpsp_j30_25_5 | min | 27 - 3s - 995n - P | 27 - 1s - 2470n - P
mrcpsp_j30_29_3 | min | 32 - 22s - 24110n - U | 32 - 57s - 166551n - U 
mrcpsp_j30_33_5 | min | 48 - 52s - 7150n - P | 48 - 5s - 9829n - P
mrcpsp_j30_41_8 | min | 30 - 42s - 5822n - P | 30 - 4s - 7984n - P 
mrcpsp_j30_54_2 | min | 28 - 12s - 9537n - U | 28 - 23s - 75877n - P


**Edit 1** : I added the number of nodes  (XXXn) and the termination (P for proven, U for unproven). 
 Now, it seems that `base-lcg` is able to find solutions exploring fewer nodes, but since the decomposition can be big, it also wastes time propagating (see for instance, `mrcpsp_j30_25_5` or `mrcpsp_j30_33_5`).
This, also, mitigates `mrcpsp_j30_54_2` because `base-lcg` is not able to prove the bound, unlike `new-lcg`.

We can conclude that the decomposition provides better explanations (probably, thanks to additional variables introduced by the decomposition) - which is, as far as I known, not surprising.   
On the other hand, the global version of cumulative provides better propagation performances but suffer from weaker explanations.

Question is: can we live with that? 
<details><summary>My opinion</summary>Yes, we can</details>

cprudhom commented:
Can we consider this PR as ready to be merged ?

ArthurGodet commented:
I didn't see your edits. So it seems that mrcpsp_j30_29_3 is the only instance that is better treated by base_lcg, and since the best found solution is not proven to be optimal, I'd say that it is not that important (especially compared to the better results we get wih new_lcg for the other instances). I agree that the explanations of new_lcg seem weaker, but my first impression would be that it is not surprising (even if I can't really explain why).

For me, even if it can still be improved (especially by better detecting when to apply the OverloadChecking or not, to improve performance), I consider that the PR can be merged. Don't forget to **_squash and merge_** the PR as the git history of the branch is awful.
